src/app/icd/client/CheckInHandler.cpp

```diff
@@ -109,8 +109,21 @@ CHIP_ERROR CheckInHandler::OnMessageReceived(Messaging::ExchangeContext * ec, co
 
     if (refreshKey)
     {
-        // TODO: A new CASE session should be established to re-register the client using a new key. The registration will happen in
-        // CASE session callback
+        RefreshKeySender * refreshKeySender = mpCheckInDelegate->OnKeyRefreshNeeded(clientInfo, mpICDClientStorage);
+        if (refreshKeySender == nullptr)
+        {
+            ChipLogError(ICD, "Key Refresh failed for node ID:" ChipLogFormatScopedNodeId,
+                         ChipLogValueScopedNodeId(clientInfo.peer_node));
+            mpCheckInDelegate->OnKeyRefreshDone(refreshKeySender, CHIP_ERROR_NO_MEMORY);
```

Comment:
OK, so the fact that OnKeyRefreshDone will get called with null was _very much not documented in the API_.

If that's the contract, that we call OnKeyRefreshDone even if OnKeyRefreshNeeded, then that needs to be very clearly documented.  That said, it doesn't seem useful to do that....

Reply:
Understood. I've added your suggestion in the description for OnKeyRefreshNeeded and removed OnKeyRefreshDone from the null use-case.